File: problem20.cpp
#include <map>
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        // Stack for opening brackets
        stack<int> openStack;

        // Map for mapping each opening brackets to closing brackets
        map<char, char> bracket;
        bracket['('] = ')';
        bracket['{'] = '}';
        bracket['['] = ']';
        // Loop throug all the char in s 
        for(int i = 0; i < s.length(); i++){
            // If it is opening brackets, add to stack
            if(bracket.count(s[i])){
                openStack.push(s[i]);
            }
            // If it is closing bracket, compare with opening bracket from the stack
            else if(s[i] == ')'){
                if(openStack.empty()){
                    return false;
                }
                char open = openStack.top();
                if(bracket[open] != s[i]){
                    return false;
                }
                openStack.pop();
            }
            else if(s[i] == '}'){
                if(openStack.empty()){
                    return false;
                }
                char open = openStack.top();
                if(bracket[open] != s[i]){
                    return false;
                }
                openStack.pop();
            }
            else if(s[i] == ']'){
                if(openStack.empty()){
                    return false;
                }
                char open = openStack.top();
                if(bracket[open] != s[i]){
                    return false;
                }
                openStack.pop();
            }
        }
        if(openStack.empty()){
            return true;
        }
        else{
            return false;
        }
    }
};

File: problem20.py
class Solution:
    def isValid(self, s: str):
        # Priority is important to close in order
        # Some case no need to extra calculation
        # If there is odd number of string there is at least one that do not have 
        # close or open bracket
        if len(s) % 2 == 1:
            return False
        # Create stack for charecter
        # To be true it need to be empty
        stack = []
        for i in s:
            # Go througb stack from the top
            stack.append(i)
        opening = 0
        closing = 0
        index = 0
        while index < len(s) and len(stack) > 0:
            # If first item in stack is opening bracket 
            # We push them into top of stack
            # If it is close item, we match with top of stack
            # If they do not match return false
            if stack[0] == "(" or stack[0] == "{" or stack[0] == "[":
                stack.append(stack[0])
                stack.pop(0)
                opening += 1
            elif stack[0] == ")":
                if stack.pop() != "(":
                    return False
                else:
                    stack.pop(0)
                    closing += 1
            elif stack[0] == "}":
                if stack.pop() != "{":
                    return False
                else:
                    stack.pop(0)
                    closing += 1
            elif stack[0] == "]":
                if stack.pop() != "[":
                    return False
                else:
                    stack.pop(0)     
                    closing += 1
            index += 1   
        # We only need to go through string one more time is enough
        if len(stack) == 0 and closing == opening:
            return True
        else:
            return False
        